DESIGN.md

```diff
@@ -0,0 +1,416 @@
+leds you defer choices between compression, checksumming, and 7-bit
+cleanliness algorithms until after a system is live and under load.
+
+As of 2017-09-13, OAK is used by ALI for volatile caches in Redis, and
+for durable Correspondence bodies in S3.
+
+Author:
+- [jhw@prosperworks.com](mailto:jhw@prosperworks.com)
+
+Advisors:
+- [marshall@prosperworks.com](mailto:marshall@prosperworks.com)
+- [gerald@prosperworks.com](mailto:gerald@prosperworks.com)
+- [kelly@prosperworks.com](mailto:kelly@prosperworks.com)
+- [neil@prosperworks.com](mailto:neil@prosperworks.com)
+
+Here is a sneak preview of some OAK strings:
+```
+$ echo 'HelloWorld!' | bin/oak.rb --format none
+oak_3CNN_1336599037_18_F1SU11_HelloWorld!_ok
+
+$ echo 'HelloWorld!' | bin/oak.rb
+oak_3CNB_1336599037_24_RjFTVTExX0hlbGxvV29ybGQh_ok
+
+$ echo 'HelloWorld!' | bin/oak.rb --compression lz4 --force
+oak_3C4B_1336599037_28_EvADRjFTVTExX0hlbGxvV29ybGQh_ok
+
+$ echo 'HelloWorld!' | bin/oak.rb | bin/oak.rb --mode decode-lines
+HelloWorld!
+```
+
+## OAK Version History
+
+OAK has been live for 16 months by the time this arch document was
+prepared retroactively.
+
+* [https://github.com/ProsperWorks/ALI/pull/1245](https://github.com/ProsperWorks/ALI/pull/1245) **oak**
+    * Merged to major_2016_04_dragonfruit.
+    * Initial implementation. Not integrated or active.
+    * Version oak_1
+* [https://github.com/ProsperWorks/ALI/pull/1350](https://github.com/ProsperWorks/ALI/pull/1350) **oak-in-summary_accessor**
+    * Merged to major_2016_04_dragonfruit.
+    * Reworked SummaryAccessor with ENV flag to switch to OAK serialization.
+    * Exposure to live data revealed some issues missed in the lab.
+* [https://github.com/ProsperWorks/ALI/pull/1631](https://github.com/ProsperWorks/ALI/pull/1631) **oak-remove-json-and-yaml**
+    * Merged to major_2016_06_goat.
+    * Simplified down to only the one serialization algorithm, "FRIZZY".
+    * Version oak_2
+* [https://github.com/ProsperWorks/ALI/pull/1618](https://github.com/ProsperWorks/ALI/pull/1618) **fix-oak-utf8**
+    * Merged in major_2016_06_goat.
+    * Fixed issues uncovered in SummaryAccessor ramp.
+    * Version oak_3
+* [https://github.com/ProsperWorks/ALI/pull/1655](https://github.com/ProsperWorks/ALI/pull/1655) **volatile_cache_accessor**
+    * Merged in major_2016_06_goat.
+    * Introduced RedisCache.  Later PRs use RedisCache for
+        * SummaryAccessor (OAK)
+        * RussianDoll caches (Marshal+OAK)
+        * S3 cache (OAK)
+        * COMPANY_ID_CACHE (JSON)
+    * Commits to OAK for volatile use cases.
+* [https://github.com/ProsperWorks/ALI/pull/1757](https://github.com/ProsperWorks/ALI/pull/1757) **oak-woe-2016-06-30**
+    * Merged in major_2016_07_hotpocket.
+    * Fixed a regexp which had broken some Float parsing.
+* [https://github.com/ProsperWorks/ALI/pull/1724](https://github.com/ProsperWorks/ALI/pull/1724) **correspondences-in-s3-fixes**
```

Comment:
A non-Prosperworks employee will not have access to the `ALI` project in Github, certainly